### ISSUE TITLE
Generic Callable[[T], Any] cannot be passed on to another generic function

### ISSUE THREAD

**Issue by kitizz** [microsoft/pylance-release#4219]
## Environment data

-   Language Server version: 2023.4.11 (pyright 64a33975)
-   OS and version: MacOS 13.0
-   Python version (& distribution if applicable, e.g. Anaconda): 3.11 (homebrew)
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: basic

## Code Snippet

```python
from typing import TypeVar
from collections.abc import Callable


T = TypeVar("T")

def foo(func: Callable[[T], bool], b: T) -> T:
    # d: dict[T, int] = {}
    return bar(func, {})


M = TypeVar("M")

def bar(func: Callable[[M], bool], b: dict[M, int]):
    ...
```

## Repro Steps

1. Define two generic functions that take a generic `Callable[[T], Any]` as an input.
2. The second function should also take a `dict[T, Any]` or `dict[Any, T]` as input.
3. The first function should try to pass the callable to the second function.
4. Typing error, "Type M@bar cannot be assigned to type T@Foo"

Error is still raised under all following tweaks to the above code:
- `T` is reused for both functions
- Dict input is fully annotated before passing in (see commented out line)
- Input args are re-ordered
- Callable is updated to take multiple other args in any order

Error is no longer raised when:
- `dict` is changed to list or tuple or just the generic TypeVar

## Expected behavior

No typing errors.

## Actual behavior

Typing problem is flagged:
```
Argument of type "(T@foo) -> bool" cannot be assigned to parameter "func" of type "(M@bar) -> bool" in function "bar"
  Type "(T@foo) -> bool" cannot be assigned to type "(M@bar) -> bool"
    Parameter 1: type "M@bar" cannot be assigned to type "T@foo"
      Type "M@bar" cannot be assigned to type "T@foo"
```

## Logs

```
[Info  - 13:32:10] (43794) Pylance language server 2023.4.11 (pyright 64a33975) starting
[Info  - 13:32:10] (43794) Server root directory: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist
[Info  - 13:32:10] (43794) Auto-indent enabled
[Info  - 13:32:12] (43794) Starting service instance "<default>"
[Info  - 13:32:12] (43794) Background analysis(1) root directory: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist
[Info  - 13:32:12] (43794) Background analysis(1) started
(43794) No pyproject.toml file found.
[Info  - 13:32:12] (43794) Setting pythonPath for service "<default>": "/opt/homebrew/bin/python3.11"
[Warn  - 13:32:12] (43794) stubPath typings is not a valid directory.
[Info  - 13:32:12] (43794) Assuming Python version 3.11
(43794) Assuming Python platform Darwin
[Info  - 13:32:12] (43794) Search paths for <default>
[Info  - 13:32:12] (43794)   /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib
[Info  - 13:32:12] (43794)   typings
[Info  - 13:32:12] (43794)   /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stubs/...
[Info  - 13:32:12] (43794)   /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/bundled/stubs
[Info  - 13:32:12] (43794)   /opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11
[Info  - 13:32:12] (43794)   /opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/lib-dynload
[Info  - 13:32:12] (43794)   /opt/homebrew/lib/python3.11/site-packages
[Info  - 13:32:12] (43794) Adding fs watcher for library directories:
 /opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11
/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/lib-dynload
/opt/homebrew/lib/python3.11/site-packages
(43794) Searching for source files
[Info  - 13:32:12] (43794) No source files found.
(43794) [IDX(FG)] index libraries  (index) ...
(43794) [IDX(FG)]   read stdlib indices (13ms)
(43794) [IDX(FG)] index libraries  (index) [succeed] (14ms)
(43794) Background analysis message: setConfigOptions
(43794) Background analysis message: setImportResolver
(43794) Background analysis message: ensurePartialStubPackages
(43794) Background analysis message: setConfigOptions
(43794) Background analysis message: setTrackedFiles
(43794) Background analysis message: markAllFilesDirty
(43794) Background analysis message: setFileOpened
(43794) [FG] parsing: Untitled-1 (18ms)
(43794) [FG] parsing: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 2ms] (66ms)
(43794) [FG] binding: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/builtins.pyi (17ms)
(43794) [FG] binding: Untitled-1 (0ms)
(43794) Background inlayHint message
(43794) [BG(1)] parsing: Untitled-1 (21ms)
(43794) [BG(1)] parsing: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 2ms] (62ms)
(43794) [BG(1)] binding: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/builtins.pyi (19ms)
(43794) [BG(1)] binding: Untitled-1 (0ms)
(43794) Background analysis message: getSemanticTokens range
(43794) [BG(1)] getSemanticTokens range 0:0 - 0:0 at Untitled-1 (1ms)
(43794) Background analysis message: setFileOpened
(43794) Background analysis message: markFilesDirty
(43794) Background analysis message: getSemanticTokens range
(43794) [BG(1)] getSemanticTokens range 0:0 - 14:7 at Untitled-1 ...
(43794) [BG(1)]   parsing: Untitled-1 (1ms)
(43794) [BG(1)]   binding: Untitled-1 (0ms)
(43794) [BG(1)]   parsing: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/typing.pyi [fs read 1ms] (15ms)
(43794) [BG(1)]   binding: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/typing.pyi (3ms)
(43794) [BG(1)]   parsing: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/typing_extensions.pyi [fs read 1ms] (3ms)
(43794) [BG(1)]   binding: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/typing_extensions.pyi (4ms)
(43794) [BG(1)]   parsing: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi [fs read 0ms] (11ms)
(43794) [BG(1)]   binding: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi (1ms)
(43794) [BG(1)]   parsing: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/types.pyi [fs read 0ms] (4ms)
(43794) [BG(1)]   binding: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/types.pyi (2ms)
(43794) [BG(1)]   parsing: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/abc.pyi [fs read 0ms] (1ms)
(43794) [BG(1)]   binding: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/abc.pyi (0ms)
(43794) [BG(1)]   parsing: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/collections/abc.pyi [fs read 0ms] (0ms)
(43794) [BG(1)]   binding: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/collections/abc.pyi ...
(43794) [BG(1)]     parsing: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/_collections_abc.pyi [fs read 0ms] (1ms)
(43794) [BG(1)]     binding: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/_collections_abc.pyi (0ms)
(43794) [BG(1)]   binding: /Users/kit/.vscode/extensions/ms-python.vscode-pylance-2023.4.11/dist/typeshed-fallback/stdlib/collections/abc.pyi (1ms)
(43794) [BG(1)] getSemanticTokens range 0:0 - 14:7 at Untitled-1 (72ms)
(43794) [FG] parsing: Untitled-1 (1ms)
(43794) [FG] binding: Untitled-1 (1ms)
(43794) Background analysis message: analyze
(43794) [BG(1)] analyzing: Untitled-1 ...
(43794) [BG(1)]   checking: Untitled-1 ...
(43794) [BG(1)]     parsing: /opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/typing.py [fs read 1ms] (31ms)
(43794) [BG(1)]     binding: /opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/typing.py (9ms)
(43794) [BG(1)]     parsing: /opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/collections/abc.py [fs read 1ms] (1ms)
(43794) [BG(1)]     binding: /opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/collections/abc.py (0ms)
(43794) [BG(1)]   checking: Untitled-1 (54ms)
(43794) [BG(1)] analyzing: Untitled-1 (54ms)
(43794) Background analysis message: getDiagnosticsForRange
(43794) Background analysis message: resumeAnalysis
(43794) Background analysis message: getDiagnosticsForRange
(43794) Background analysis message: getDiagnosticsForRange
(43794) Background analysis message: getDiagnosticsForRange
(43794) IntelliCode model /Users/kit/.vscode/extensions/visualstudioexptteam.vscodeintellicode-1.2.30/dist/bundledModels/python_intellisense-members-lstm-pylance
(43794) Background inlayHint message
(43794) Loading ONNX runtime...
(43794) Loaded ONNX runtime. Creating IntelliCode session...
2023-04-10 13:32:14.540871 [I:onnxruntime:, inference_session.cc:263 operator()] Flush-to-zero and denormal-as-zero are off
2023-04-10 13:32:14.541337 [I:onnxruntime:, inference_session.cc:271 ConstructorCommon] Creating and using per session threadpools since use_per_session_threads_ is true
2023-04-10 13:32:14.541420 [I:onnxruntime:, inference_session.cc:292 ConstructorCommon] Dynamic block base set to 0
2023-04-10 13:32:14.547186 [I:onnxruntime:, inference_session.cc:1222 Initialize] Initializing session.
2023-04-10 13:32:14.547262 [I:onnxruntime:, inference_session.cc:1259 Initialize] Adding default CPU execution provider.
2023-04-10 13:32:14.548639 [I:onnxruntime:, reshape_fusion.cc:42 ApplyImpl] Total fused reshape node count: 0
2023-04-10 13:32:14.549160 [I:onnxruntime:, graph.cc:3553 CleanUnusedInitializersAndNodeArgs] Removing NodeArg 'rnn/transpose'. It is no longer used by any node.
2023-04-10 13:32:14.549195 [I:onnxruntime:, graph.cc:3553 CleanUnusedInitializersAndNodeArgs] Removing NodeArg 'Gather_out0'. It is no longer used by any node.
2023-04-10 13:32:14.549226 [I:onnxruntime:, graph.cc:3553 CleanUnusedInitializersAndNodeArgs] Removing NodeArg 'Transpose_out0'. It is no longer used by any node.
2023-04-10 13:32:14.549377 [I:onnxruntime:, reshape_fusion.cc:42 ApplyImpl] Total fused reshape node count: 0
2023-04-10 13:32:14.552187 [V:onnxruntime:, session_state.cc:1010 VerifyEachNodeIsAssignedToAnEp] Node placements
2023-04-10 13:32:14.552237 [V:onnxruntime:, session_state.cc:1013 VerifyEachNodeIsAssignedToAnEp]  All nodes placed on [CPUExecutionProvider]. Number of nodes: 59
2023-04-10 13:32:14.552490 [V:onnxruntime:, session_state.cc:66 CreateGraphInfo] SaveMLValueNameIndexMapping
2023-04-10 13:32:14.552570 [V:onnxruntime:, session_state.cc:112 CreateGraphInfo] Done saving OrtValue mappings.
2023-04-10 13:32:14.552828 [I:onnxruntime:, session_state_utils.cc:199 SaveInitializedTensors] Saving initialized tensors.
2023-04-10 13:32:14.557036 [I:onnxruntime:, session_state_utils.cc:342 SaveInitializedTensors] Done saving initialized tensors
2023-04-10 13:32:14.557488 [I:onnxruntime:, inference_session.cc:1488 Initialize] Session successfully initialized.
(43794) Created IntelliCode session.
(43794) Initialize deeplearning succeeded
(43794) Background analysis message: getSemanticTokens range
(43794) [BG(1)] getSemanticTokens range 0:0 - 14:7 at Untitled-1 (2ms)
(43794) Background inlayHint message
(43794) Background analysis message: getSemanticTokens range
(43794) [BG(1)] getSemanticTokens range 0:0 - 14:7 at Untitled-1 (2ms)
(43794) Background inlayHint message
(43794) Background inlayHint message
(43794) Background inlayHint message
(43794) Background inlayHint message
(43794) Background analysis message: getSemanticTokens range
(43794) [BG(1)] getSemanticTokens range 0:0 - 14:7 at Untitled-1 (2ms)
(43794) Background analysis message: getDiagnosticsForRange
(43794) Background analysis message: getDiagnosticsForRange

```

**Comment by erictraut** [comment]
Yes, I agree this is a bug. Thanks for reporting it.

The type evaluator is not properly handling bidirectional type inference for an empty dictionary statement in this situation. I'll need to investigate in more detail.

**Comment by kitizz** [comment]
Some more information. If I change the definition of `T` to:
```python
T = TypeVar("T", int, float)
```
It seems to be ok. But `bound=...` still produces an error

**Comment by bschnurr** [comment]
This issue has been fixed in prerelease version 2023.4.21, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023421-12-April-2023-PreRelease)
